src/selectors/dataset.js

```diff
@@ -14,16 +14,91 @@ export const datasetSelector = (config, state, instance) => {
-  const instanceData = baseState.getIn(['instances', instance, 'data'])
+  let instanceData = baseState.getIn(['instances', instance, 'data'])
   const rawData = baseState.get('raw')
 
   // If instanceData is not created yet, return empty Map
   if(!instanceData) {
     return Map({})
   }
 
+  // Remove the pending.create items from the instance
+  instanceData = instanceData.filter(uid => baseState.getIn(['pending', 'create'], List()).indexOf(uid) === -1)
+
+  // Re-compute data
+  const computedData = instanceData.map(uid => {
+    let item = rawData.get(uid)
+
+    item = item.withMutations(map => {
+      // Check pending.update and add the pendingUpdate prop if necessary
+      const pendingUpdate = baseState.getIn(['pending', 'update'], List()).indexOf(uid) > -1
+      if (pendingUpdate) {
+        map.set('pendingUpdate', true)
+      }
+      // Check pending.remove and add the pendingRemove prop if necessary
+      const pendingRemove = baseState.getIn(['pending', 'remove'], List()).indexOf(uid) > -1
+      if (pendingRemove) {
+        map.set('pendingRemove', true)
+      }
+    })
+
+    return item
+  })
+
+  // Gather additional data
+  const additionalData = baseState.getIn(['instances', instance, 'additionalData']) || Map({})
+
+  return additionalData.merge(Map({ data: computedData }))
+}
+
+/**
+ * The selector for computing derived datasets with pending data
+ * @param  {Object} config - selector configuration
+ * @param  {Immutable.Map} state - current reducer state
+ * @param  {Object} instance - instance key
+ *
+ * @return {Immutable.Map} - computed data
+ */
+export const datasetOptimisticSelector = (config, state, instance) => {
+
+  // Get redux-clerk data from store
+  const baseState = config.baseSelector(state)
+
+  // Get instance array
+  let instanceData = baseState.getIn(['instances', instance, 'data'])
+
+  let rawData = baseState.get('raw')
+
+  // If instanceData is not created yet, return empty Map
+  if(!instanceData) {
+    return Map({})
+  }
+
+  // Remove the pending.remove items
+  instanceData = instanceData.filter(uid => baseState.getIn(['pending', 'remove'], List()).indexOf(uid) === -1)
+
+  // Merge with pendingRaw to account for updated items
+  rawData = rawData.mergeDeep(baseState.get('pendingRaw'))
+
   // Re-compute data
-  const computedData = instanceData.map(i => rawData.get(i))
+  const computedData = instanceData.map(uid => {
+    let item = rawData.get(uid)
+
+    item = item.withMutations(map => {
+      // Check pending.create and add the pendingCreate prop if necessary
+      const pendingCreate = baseState.getIn(['pending', 'create'], List()).indexOf(uid) > -1
+      if (pendingCreate) {
+        map.set('pendingCreate', true)
+      }
+      // Check pending.update and add the pendingUpdate prop if necessary
+      const pendingUpdate = baseState.getIn(['pending', 'update'], List()).indexOf(uid) > -1
+      if (pendingUpdate) {
+        map.set('pendingUpdate', true)
+      }
```

Comment:
Get `baseState.getIn(['pending', 'update'], List())` and `baseState.getIn(['pending', 'remove'], List())` outside of loop.